pkg/authentication/authentication.go

```diff
@@ -0,0 +1,25 @@
+package authentication
+
+import (
+	"context"
+)
+
+const (
+	// SchemeBearer defines a Bearer token scheme.
+	SchemeBearer = "Bearer"
+)
+
+// Credentials contains information about an authentication request.
+type Credentials struct {
+	// Scheme contains information about the type of credentials being used.
+	Scheme string
+	// Token contains information the actual credential.
+	Token string
+}
```

Comment:
I'd like to have a conversation about using this generic approach to authenticating requests as I feel it might restrict us in the future and provide poor ergonomics for the end user.

We're essentially using the `Scheme` field of this struct to allow our single `VerifyCredentials` authentication method to multiplex to an internal function that performs the specific required form of authentication (e.g. bearer token, basic authentication, access token, etc.).

This imposes a non-negligible complexity on the client using the library, which I will get into, and makes it hard or costly to add and remove fields to this struct as the potential impact to **all** forms of authentication have to be considered.

As an example, a client wanting to validate a bearer token would have to do the following

```go
token := GetTokenSomehow()
creds := &Credentials{
  Scheme: authorization.Bearer
  Token: token
}
// Assume ctx was already defined
authSvc.VerifyCredentials(ctx, creds)
```

The `auth` service or library then receives the credentials, `switch`es on the `Scheme` field and calls an internal function that performs the bearer token validation.

Now, in theory the benefit this gives the client is that they can change the credential type with a simple value, 

```go
token := GetTokenSomehow()
creds := &Credentials{
  // This now uses private token authentication!
  Scheme: authorization.PrivateToken
  Token: token
}
// Assume ctx was already defined
authSvc.VerifyCredentials(ctx, creds)
```

This benefit however only holds if all authorization schemes have the same fields. If we ever want to add support for basic HTTP auth (user and password), we have two options:

1. Adding `User` and `Password` fields to the `Credentials` struct and adding a bunch of documentation to explain when to use the different sets of fields in the struct (you won't need to fill in the `Token` field if you're using basic auth and vice-versa for bearer token auth).
2. Encode the user and password into a single field using base64 (like it usually happens in the browser).

If the client only wants to use one form of authentication, they need to be cognizant of the fields they need to fill in when passing credentials. If they want to support multiple forms of authentication they will have to select the auth type, and then conditionally fill in the `Credentials` fields based on the type of auth they're about to perform.

Option 2 is technically viable **all** authentication methods as any set of data can be compressed into a single string value. As an example, you can define a struct specific to an authentication scheme, serialize it to JSON and then have the auth library implementation deserialize the JSON to get the struct back. The issue with this is that it isn't exactly performant, and it is somewhat inconvenient for the user as it adds one more boilerplate step.

```go
credType := GetCredType()
creds := &Credentials{
  Scheme: authorization.Basic
  // ToToken in this case compresses User and Password into a single base64 value that can be 
  // understood by the auth library implementation.
  Token: authorization.BasicCredentials{
    User: "test",
    Password: "test",
  }.toToken() 
}
// Assume ctx was already defined
authSvc.VerifyCredentials(ctx, creds)
```

If we had a method for each type of authentication as such

```go
type Authentication interface {
  // Naming can definitely be improved
  ValidateBearerToken(ctx context.Context, token string) error
  ValidatePrivateToken(ctx context.Context, token string) error
  ValidateBasicAuth(ctx context.Context, user string, password string) error
  [...]
}
```

then a lot of this complexity and boilerplate would simply go away.

```go
// Assume ctx was already defined
authSvc.VerifyBearerToken(ctx, GetTokenSomehow())
```

1. There would be no need for the end client to know how to structure the data for a specific authorization scheme as it would be defined in the method signature. This has the extra benefit that IDEs would help with autocomplete.
2. It would reduce the complexity for us as we would not have to consider the needs multiple authorization schemes in a single struct. 
3. It would remove the need to serialize and deserialize more complex data required by some authorization schemes.
4. It would remove the need to multiplex off the `Scheme` field in the auth library implementation, making it more efficient.

Reply:
Thanks for raising this comment, this is certainly valuable. My original plan was to support different types of Schemes in a single Auth service, and compose it with multiple `Authentication` implementations for an Auth provider. Meaning that if Auth0 supports `Bearer`, `API Key` and `Private-Tokens`, 3 `Authentication` implementations will be added to the `Auth0`  service, and will internally multiplex  based on the `Scheme`. Adding support for another Auth provider such as Google Identity Platform would consist of essentially composing different `Authentication` implementations as needed.

![image](https://user-images.githubusercontent.com/5170276/234590358-52afced0-aab8-47fb-83e7-fb76bf9cffce.png)

Reply:
We can expose an interface like the one you described in your comment in the `Auth` interface, something like:

```go
type Auth interface {
  // Naming can definitely be improved
  ValidateBearerToken(ctx context.Context, token string) error
  ValidatePrivateToken(ctx context.Context, token string) error
  ValidateBasicAuth(ctx context.Context, user string, password string) error
  [...]
}
```

But that's another layer of abstraction that sits on top of this PR, and can be added in another PR.

Reply:
> My original plan was to support different types of Schemes in a single Auth service, and compose it with multiple Authentication implementations for an Auth provider. Meaning that if Auth0 supports Bearer, API Key and Private-Tokens, 3 Authentication implementations will be added to the Auth0 service, and will internally multiplex based on the Scheme. Adding support for another Auth provider such as Google Identity Platform would consist of essentially composing different Authentication implementations as needed.

My comment did not touch on this, refer to the overall design or how we end up implementing this on our side; but I am aligned with the initial goal of supporting different schemes. I'll get to this point later in this comment. My focus for this comment is on improving the end user experience and reducing maintenance burden for us in the future.

The main problem with having a single multiplexing function is that the data that you need to pass to each `Authenticate` call would still have to vary depending on the internal authentication scheme. Wrapping the information in a `Credentials` struct only adds more boilerplate without solving the problem; you still have to specify the `Scheme`, which is essentially identical to calling a different function, and you still need to provide the scheme-specific information through the `Token` field (or otherwise). 

I believe our stance should always be to make the most common use case as simple as possible to setup (minimize the number of calls, ideally with no config and sensible defaults if possible), but still allow the end user to implement exceptions if needed. 

In this case, in the rare case where an end user wants to use multiple authentication schemes based on a configuration, say a `string`, they can implement this on their side. I consider this use case to be an exception and not the norm, and as such the end user will have to add code to handle the exceptional situation. It is up to the them to perform the multiplexing on their side and call the appropriate function. 

```go
authNType := GetAuthNType()
switch authNType {
case BEARER:
  authSvc.VerifyBearerToken(ctx, getTokenSomehow())
case PRIVATE:
  authSvc.VerifyBearerToken(ctx, getTokenSomehow())
case BASIC:
  authSvc.VerifyBasicAuth(ctx, getUser(), getPW())
}

```

With the `VerifyCredentials` approach this `switch`  would have to be on our side of the implementation and be run everytime a user verifies credentials. In this proposal it ends up being delegated to the user and is only required in exceptional cases where the user wants to support multiple schemes. This provides saner defaults, less boilerplate for the end user in the more general use of the library, and less headaches for us as we don't have to support multiple use cases with a single function.

In short, with this proposal you get the same result as with `VerifyCredentials`, only you remove an extra step on both ends: the user does not need to configure the data to send and the implementation does not need to multiplex.

As a final note, given that the authentication schemes are typically standards with very well defined inputs and outputs, and that the different providers will have to adhere those standards, it is very likely that all providers will handle the same inputs and outputs for each authentication scheme, and therefore it makes sense for the interface function signature to define the set of inputs for all implementations.

### Supporting different schemes

> We can expose a set of methods like the ones you described in your comment, but in the Auth interface instead; something like: [...]

Relegating the authentication API problems to the `Auth` interface is probably not the way to go. 

In general, having a higher level layer expose an endpoint to expose a functionality that can be entirely performed by a single lower layer component does not make sense as it is the responsibility of the lower layer component to design the API to expose its functionality. In this particular case, I believe the `Auth` interface should just compose all of the methods that are expected of our auth library, and this includes both `Authentication` and `Authorization`.

Here's a quick proposal for the `Auth` library:

![image](https://user-images.githubusercontent.com/483619/234687292-bb1ff043-ffc4-443b-8702-e10eb7e91b90.png)

This design does make supporting different authorization schemes with different providers require some extra steps because the providers are defined at the interface level. Having multiple providers for an `Authentication`/`Authorization` implementation should be considered an exception and up to the end user to implement. This special use case can be implemented like so:

```go
// MyAuth implements `Authorization` with multiple providers
type MyAuth struct {
  // Auth0AuthN is the Auth0 `Authorization` implementation
  Auth0AuthN
  // GoogleISAuthN is the Google Identity Service `Authorization` implementation
  GoogleISAuthN
}

// Bearer scheme provided by Auth0
func (ma *MyAuth) VerifyBearerToken error { return ma.Auth0AuthN.VerifyBearerToken }

// Other schemes provided by Google IS
func (ma *MyAuth) VerifyPrivateToken error { return ma.GoogleISAuthN.VerifyPrivateToken }
func (ma *MyAuth) VerifyBasicAuth error { return ma.GoogleISAuthN.VerifyBasicAuth }
```

```go
type auth struct {
  Authorization
  Authentication
}

func NewAuth(authorization Authorization, authentication Authentication) Auth { [...] }
```

We can opt to define an interface for each `Authentication` scheme and allow partial interface implementation by providers, but I don't think this will be a common use case, and having each implementation return a `NotImplemented` or `NotSupported` error is fine for now. We can revisit this assumption if things change.

### Usage example

To try to illustrate my point, I'd like to write a simple demo of what using our library looks like.

```go
import (
  "github.com/gazebo-web/auth/auth"
  "github.com/gazebo-web/auth/authentication"
  "github.com/gazebo-web/auth/authorization"
  "github.com/gazebo-web/auth/middleware"
)

func main() {
  // Here we pick our Authentication implementation (and therefore provider(s))
  authN := authentication.NewMyAuth(amazing, params)

  // Here we pick our Authorization implementation (and therefore provider)
  authZ := authorization.NewSpiceDB(very, cool, params)

  // This is where everything comes together into a nice, single bundle
  auth := auth.NewAuth(authN, authZ)

  // Prepare a BearerToken middleware with the configured auth clients
  bearerMW := middlewares.BearerToken(auth)

  // Tie auth into endpoints and services from here on
  [...]
}
```